cmd/generic-handlers.go

```
@@ -576,3 +577,59 @@ func setCriticalErrorHandler(h http.Handler) http.Handler {
 		h.ServeHTTP(w, r)
 	})
 }
+
+// guessIsMultipartUploadReq - returns true if incoming request is multipart upload call other than CreateMultipartUpload
+func guessIsMultipartUploadReq(r *http.Request) bool {
+	if r == nil {
+		return false
+	}
+	for k := range r.Form {
+		if strings.EqualFold(k, "uploadId") {
+			return true
+		}
+	}
+	return false
+}
+
+// setUploadForwardingHandler middleware forwards multiparts requests
+// in a site replication setup to peer that initiated the upload
+func setUploadForwardingHandler(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !globalSiteReplicationSys.isEnabled() ||
+			guessIsHealthCheckReq(r) || guessIsMetricsReq(r) ||
+			guessIsRPCReq(r) || guessIsLoginSTSReq(r) || isAdminReq(r) {
+			h.ServeHTTP(w, r)
+			return
+		}
+		bucket, object := request2BucketObjectName(r)
+		if bucket != "" && object != "" && guessIsMultipartUploadReq(r) {
+			uploadID := r.Form.Get(xhttp.UploadID)
+			pi, local := globalSiteReplicationSys.getPeerForUpload(r.Context(), uploadID)
+			if local {
+				h.ServeHTTP(w, r)
+				return
+			}
+			// forward request to peer handling this upload
+			epURL, err := url.Parse(pi.Endpoint)
+			if err != nil {
+				h.ServeHTTP(w, r)
```

```suggestion
```

Lets return an error appropriately.. or have pi.Endpoint already a parsed endpoint.